File: backend/src/voice_agent/services/llm.py
"""LLM service — GPT-4o with function calling and two-turn pattern."""

import json
from dataclasses import dataclass, field

from openai import AsyncOpenAI

from voice_agent.config import get_settings
from voice_agent.domain.banking_models import TOOL_SCHEMAS
from voice_agent.domain.prompts import build_llm_messages, build_system_prompt
from voice_agent.lib.errors import LLMError
from voice_agent.lib.logger import get_logger
from voice_agent.lib.timer import TimingResult, timer
from voice_agent.services import banking

logger = get_logger(__name__)


@dataclass
class LLMResult:
    """Result of an LLM turn."""

    text: str
    tool_calls: list[dict] = field(default_factory=list)  # [{"name": ..., "args": {...}, "result": ...}]
    language_detected: str | None = None
    timing_ms: float = 0.0
    stt_timing_ms: float = 0.0  # passed in from pipeline


@dataclass
class ToolCallResult:
    """Result of executing a single tool."""

    name: str
    args: dict
    result: str  # JSON string of result
    success: bool
    error: str | None = None


class LLMService:
    """GPT-4o with function calling — handles two-turn pattern."""

    def __init__(self, client: AsyncOpenAI | None) -> None:
        self._client = client
        self._settings = get_settings()

    async def generate(
        self,
        user_transcript: str,
        conversation_turns: list[dict],
        language_detected: str | None = None,
        stt_timing_ms: float = 0.0,
    ) -> LLMResult:
        """Generate assistant response using two-turn LLM pattern.

        Turn 1: user transcript → LLM (may fire tools)
        Turn 2 (if tools): tool results → LLM → final text
        """
        if self._client is None:
            raise LLMError("OpenAI client not configured")

        system_prompt = build_system_prompt(language_detected)
        messages = build_llm_messages(system_prompt, conversation_turns)
        messages.append({"role": "user", "content": user_transcript})

        total_timing = TimingResult()
        async with total_timing:
            # First LLM call
            async with timer() as llm_t:
                response = await self._client.chat.completions.create(
                    model="gpt-4o",
                    messages=messages,
                    tools=TOOL_SCHEMAS,
                    tool_choice="auto",
                )

            choice = response.choices[0]
            assistant_message = choice.message

            # Handle tool calls
            if assistant_message.tool_calls:
                tool_results = await self._execute_tools(assistant_message.tool_calls)

                # Second LLM call — send tool results back
                messages.append({
                    "role": "assistant",
                    "content": assistant_message.content or "",
                    "tool_calls": [
                        {"id": tc.id, "function": {"name": tc.function.name, "arguments": tc.function.arguments}}
                        for tc in assistant_message.tool_calls
                    ],
                })
                for tr in tool_results:
                    messages.append({
                        "role": "tool",
                        "tool_call_id": tr.name,  # will be fixed below
                        "content": tr.result,
                    })

                # Second call
                async with timer():
                    follow_up = await self._client.chat.completions.create(
                        model="gpt-4o",
                        messages=messages,
                    )
                final_text = follow_up.choices[0].message.content or assistant_message.content or ""
            else:
                final_text = assistant_message.content or ""
                tool_results = []

        # Serialize tool calls for response
        serialized_tool_calls = [
            {
                "name": tr.name,
                "args": tr.args,
                "result": tr.result,
            }
            for tr in tool_results
        ]

        return LLMResult(
            text=final_text,
            tool_calls=serialized_tool_calls,
            language_detected=language_detected,
            timing_ms=total_timing.elapsed_ms,
            stt_timing_ms=stt_timing_ms,
        )

    async def _execute_tools(self, tool_calls: list) -> list[ToolCallResult]:
        """Execute all tool calls and return results."""
        results = []
        for tc in tool_calls:
            func_name = tc.function.name
            try:
                args = json.loads(tc.function.arguments) if tc.function.arguments else {}
            except json.JSONDecodeError:
                args = {}

            try:
                result = banking.execute_tool(func_name, args)
                result_str = json.dumps(result, ensure_ascii=False)
                results.append(ToolCallResult(name=func_name, args=args, result=result_str, success=True))
            except Exception as e:
                logger.error("tool_execution_failed", function=func_name, error=str(e))
                error_result = {"error": str(e), "function": func_name}
                results.append(ToolCallResult(
                    name=func_name,
                    args=args,
                    result=json.dumps(error_result),
                    success=False,
                    error=str(e),
                ))

        return results


def create_llm_service(api_key: str | None) -> LLMService | None:
    """Factory: return LLMService if API key present, else None."""
    if not api_key:
        logger.warning("OPENAI_API_KEY not set — LLM unavailable")
        return None
    return LLMService(AsyncOpenAI(api_key=api_key))

File: backend/src/services/llm.ts
import OpenAI from 'openai'
import { LLMError } from '../lib/errors.js'
import { logger } from '../lib/logger.js'
import { TOOL_SCHEMA_JSON } from '../domain/banking-models.js'
import { executeTool } from './banking.js'
import type { Turn } from '../domain/conversation.js'
import type { ChatCompletionMessageParam } from 'openai/resources'

export interface LLMResult {
  text: string
  toolCalls: Array<{ name: string; args: Record<string, unknown>; result: string }>
  languageDetected: string | null
  timingMs: number
}

function buildSystemPrompt(languageDetected: string | null): string {
  if (languageDetected === 'ar') {
    return `أنت مساعد مصرفي صوتي عربي ودي. تتحدث فقط العربية الفصحى السليمة.

مهمتك:
- الإجابة على أسئلة العملاء حول حساباتهم المصرفية
- عرض أرصدة الحسابات والحركات الأخيرة
- المساعدة في block/unblock البطاقات
- تقديم شكوى على معاملة
- التواصل بوضوح وبساطة

القواعد:
- أجب دائمًا بالعربية الفصحى
- لا تخترع معلومات — استخدم الأدوات المتاحة فقط
- إذا لم تفهم، اطلب التكرار بلباقة
- كن ودودًا واحترافيًا`
  }
  return `You are a friendly Arabic bilingual voice banking assistant.

Your role:
- Answer customer questions about their bank accounts
- Show account balances and recent transactions
- Help block/unblock cards
- File disputes on transactions
- Communicate clearly and simply

Rules:
- Respond in the same language the customer used (Arabic or English)
- Never invent information — use available tools only
- If you don't understand, ask politely to repeat
- Be friendly and professional`
}

function buildMessages(
  systemPrompt: string,
  conversationTurns: Turn[],
): ChatCompletionMessageParam[] {
  const messages: ChatCompletionMessageParam[] = [
    { role: 'system', content: systemPrompt },
  ]
  for (const turn of conversationTurns) {
    if (turn.role === 'user') {
      messages.push({ role: 'user', content: turn.content })
    } else if (turn.role === 'assistant') {
      const msg: ChatCompletionMessageParam = { role: 'assistant', content: turn.content }
      if (turn.tool_calls.length > 0) {
        (msg as ChatCompletionMessageParam & { tool_calls: unknown[] }).tool_calls = turn.tool_calls.map((tc, i) => ({
          id: `call_${tc.name}_${i}`,
          type: 'function' as const,
          function: { name: tc.name, arguments: JSON.stringify(tc.args) },
        }))
      }
      messages.push(msg)
    }
  }
  return messages
}

export class LLMService {
  constructor(private client: OpenAI | null) {}

  async generate(
    userTranscript: string,
    conversationTurns: Turn[],
    languageDetected: string | null = null,
    _sttTimingMs: number = 0,
  ): Promise<LLMResult> {
    if (!this.client) throw new LLMError('OpenAI client not configured')

    const systemPrompt = buildSystemPrompt(languageDetected)
    const messages = buildMessages(systemPrompt, conversationTurns)
    messages.push({ role: 'user', content: userTranscript })

    const start = performance.now()

    // First call
    const response = await this.client.chat.completions.create({
      model: 'gpt-4o',
      messages,
      tools: TOOL_SCHEMA_JSON,
      tool_choice: 'auto',
    })

    const choice = response.choices[0]
    const assistantMessage = choice.message

    let finalText = assistantMessage.content ?? ''
    const toolCallsResult: Array<{ name: string; args: Record<string, unknown>; result: string }> = []

    if (assistantMessage.tool_calls && assistantMessage.tool_calls.length > 0) {
      const toolResults: ChatCompletionMessageParam[] = []

      for (const tc of assistantMessage.tool_calls) {
        const funcName = tc.function.name
        let args: Record<string, unknown> = {}
        try { args = JSON.parse(tc.function.arguments) } catch { /* use empty */ }

        try {
          const result = executeTool(funcName, args)
          toolResults.push({
            role: 'tool',
            tool_call_id: tc.id,
            content: JSON.stringify(result, null, 2),
          })
          toolCallsResult.push({ name: funcName, args, result: JSON.stringify(result) })
        } catch (err) {
          const errMsg = err instanceof Error ? err.message : String(err)
          toolResults.push({
            role: 'tool',
            tool_call_id: tc.id,
            content: JSON.stringify({ error: errMsg }),
          })
          toolCallsResult.push({ name: funcName, args, result: JSON.stringify({ error: errMsg }) })
        }
      }

      const assistantMsgForSecondCall: ChatCompletionMessageParam = {
        role: 'assistant',
        content: assistantMessage.content ?? '',
      }
      if (assistantMessage.tool_calls.length > 0) {
        (assistantMsgForSecondCall as ChatCompletionMessageParam & { tool_calls: unknown[] }).tool_calls =
          assistantMessage.tool_calls.map((tc) => ({
            id: tc.id,
            type: 'function' as const,
            function: { name: tc.function.name, arguments: tc.function.arguments },
          }))
      }

      messages.push(assistantMsgForSecondCall)
      messages.push(...toolResults)

      // Second call
      const followUp = await this.client.chat.completions.create({
        model: 'gpt-4o',
        messages,
        tools: TOOL_SCHEMA_JSON,
      })
      finalText = followUp.choices[0].message.content ?? assistantMessage.content ?? ''
    }

    const timingMs = performance.now() - start
    return { text: finalText, toolCalls: toolCallsResult, languageDetected, timingMs }
  }
}

export function createLLMService(apiKey: string | null): LLMService | null {
  if (!apiKey) {
    logger.warn('OPENAI_API_KEY not set — LLM unavailable')
    return null
  }
  return new LLMService(new OpenAI({ apiKey }))
}
